Core/include/Acts/Material/IndexedSurfaceMaterial.hpp

```diff
@@ -0,0 +1,136 @@
+/// material maps associated to a grid structure
+///
+/// @tparam grid_type is the type of the grid used here
+/// @tparam material_accessor_type is the type of the accessor to the material
+///
+/// It is templated on the material type and a slab accessor type in order
+/// to allow it to be used in the material recording as well.
+template <typename grid_type,
+          typename material_accessor_type = MaterialSlabAccessor>
+class IndexedSurfaceMaterial : public ISurfaceMaterial {
+ public:
+  /// @brief Constructor for indexed surface material
+  /// @param material the material vector in a flat format
+  /// @param grid the index grid steering the access to the material vector
+  /// @param gcasts global casts -> casts a grid position from global
+  /// @param laccessors local accessors -> accessors to the local grid
+  /// @param transform  transform from global frame into map frame
+  IndexedSurfaceMaterial(
+      std::vector<typename material_accessor_type::material_type>&& material,
+      grid_type&& grid, const std::vector<BinningValue>& gcasts,
+      const std::vector<std::size_t>& laccessors,
+      const Transform3& transform = Transform3::Identity())
+      : m_material(std::move(material)),
+        m_grid(std::move(grid)),
+        m_globalCasts(gcasts),
+        m_localAccessors(laccessors),
+        m_transform(transform) {
+    if (gcasts.size() != grid_type::DIM) {
+      throw std::invalid_argument(
+          "The number of casts must match the grid dimension");
+    }
+
+    if (gcasts.size() != laccessors.size()) {
+      throw std::invalid_argument(
+          "The number of casts must match the number of local accessors");
+    }
+  }
+
+  /// @copydoc ISurfaceMaterial::materialSlab(const Vector2&) const
+  const MaterialSlab& materialSlab(const Vector2& lp) const final {
+    // Access via local position lookup
+    std::size_t index = m_grid.atPosition(
+        GridAccessHelpers::accessLocal<grid_type>(lp, m_localAccessors));
+    return material_accessor_type::slab(m_material[index]);
+  }
+
+  /// @copydoc ISurfaceMaterial::materialSlab(const Vector3&) const
+  const MaterialSlab& materialSlab(const Vector3& gp) const final {
+    // Access via (transformed) global position lookup
+    auto index = m_grid.atPosition(GridAccessHelpers::castPosition<grid_type>(
+        m_transform * gp, m_globalCasts));
+    return material_accessor_type::slab(m_material[index]);
+  }
+
+  /// @copydoc ISurfaceMaterial::materialSlab(std::size_t bin0, std::size_t bin1) const
+  ///
+  /// The ISurface material class expects bins in [ 0 - n ] x [ 0, m ], we will
+  /// convert into this format, but it is gonna be slow
+  const MaterialSlab& materialSlab(std::size_t bin0,
+                                   std::size_t bin1) const final {
+    // Access via bin0 and bin1
+    Acts::Vector2 lposition =
+        GridAccessHelpers::toLocal(m_grid, bin0, bin1, m_localAccessors);
```

Comment:
I thought ```toLocal``` should not be used in performance relevant code?

Reply:
Guess we need to profile this. At the end, this will happen ~10-20 times per track, so it should be ok.